src/meltano/core/elt_context.py

```diff
@@ -142,17 +147,25 @@ def __init__(
 
         self.base_output_logger = base_output_logger
         self.merge_state = merge_state
+        self.run_id = run_id
 
     @property
-    def elt_run_dir(self):  # noqa: ANN201
+    def elt_run_dir(self) -> Path | None:
         """Get the ELT run directory.
 
         Returns:
             The job dir, if a Job is provided, else None.
         """
-        if self.job:  # noqa: RET503
+        # In an EL/ELT run, there is always a job. This means the run directory
+        # is always ".meltano/run/<state-id>/<run-id>".
+        # This means the catalog can only be cached for executions that use the
+        # same state and run ID.
+        # See https://github.com/meltano/meltano/issues/2415.
+        if self.job:
             return self.project.job_dir(self.job.job_name, str(self.job.run_id))
 
+        return None  # pragma: no cover
```

Comment:
If there should always be a job, would it be okay if we replace `return None` with `raise RuntimeException` (or some other exception), and then change the return type to `Path` instead of `Path | None`?

Reply:
Yeah, I was maybe being unnecessarily cautious but it's probably fine to crash if `self.job` is null, which should never happen.

Reply:
Argh, we actually do test the `ELTContext` for cases in which the job is not set. Rather than updating the tests, I'll revert the raising of an exception here and update the comment.